xrpl/testutil/integration/runner.go

```diff
@@ -0,0 +1,139 @@
+package integration
+
+import (
+	"testing"
+
+	"github.com/Peersyst/xrpl-go/pkg/crypto"
+	"github.com/Peersyst/xrpl-go/xrpl/queries/transactions"
+	"github.com/Peersyst/xrpl-go/xrpl/transaction"
+	"github.com/Peersyst/xrpl-go/xrpl/wallet"
+	"github.com/stretchr/testify/require"
+)
+
+type Runner struct {
+	t      *testing.T
+	config *RunnerConfig
+
+	client  Client
+	wallets []*wallet.Wallet
+}
+
+// NewRunner creates a new runner. It doesn't connect to the websocket or generate wallets until Setup is called.
+// A testing.T is required to use the require package.
+func NewRunner(t *testing.T, client Client, config *RunnerConfig) *Runner {
+	return &Runner{
+		t:      t,
+		config: config,
+		client: client,
+	}
+}
+
+// Setup creates a new websocket client and generates the required number of wallets.
+// It also connects to the websocket and starts the client.
+// For every wallet, it will create a new account and fund it with the faucet.
+func (r *Runner) Setup() error {
+	if connectable, ok := r.client.(Connectable); ok {
+		err := connectable.Connect()
+		if err != nil {
+			return err
+		}
+	}
+
+	for i := 0; i < r.config.WalletCount; i++ {
+		w, err := wallet.New(crypto.ED25519())
+		if err != nil {
+			return err
+		}
+		err = r.client.FundWallet(&w)
+		if err != nil {
+			return err
+		}
+		r.wallets = append(r.wallets, &w)
+	}
+	return nil
+}
+
+// Teardown closes the websocket client.
+func (r *Runner) Teardown() error {
+	if connectable, ok := r.client.(Connectable); ok {
+		err := connectable.Disconnect()
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// TestTransaction submits a signed transaction and validates the result.
+// If validate is nil, the transaction is not validated.
+func (r *Runner) TestTransaction(flatTx *transaction.FlatTransaction, signer *wallet.Wallet) (*transactions.SubmitResponse, error) {
+	tx, hash, err := r.processTransaction(flatTx, signer)
+	if err != nil {
+		return nil, err
+	}
+
+	require.NoError(r.t, err)
+	require.Equal(r.t, tx.EngineResult, "tesSUCCESS")
```

Comment:
I think we should make `TestTransaction` more flexible and not expect `tesSUCCESS` all the time.

For example, with `DeepFreeze` we would need to write tests and a successful test would have `tecPATH_DRY` for a Payment to an account which has its trustline Deep Frozen.

So we should probably have another argument to this function to accept any error code listed here: https://xrpl.org/docs/references/protocol/transactions/transaction-results

Reply:
I agree, good catch.